skbio/stats/ordination/_cutils.pyx

```diff
@@ -0,0 +1,165 @@
+    """
+    Compute E matrix from a distance matrix, and 
+    also compute the means in the process.
+
+    Squares and divides by -2 the input elementwise. Eq. 9.20 in
+    Legendre & Legendre 1998.
+
+
+    Parameters
+    ----------
+    mat : 2D array_like
+        Distance matrix.
+    centered : 2D array_like
+        Output, E matrix. Must be pre-allocated and same shape as mat.
+        Can point to mat (i.e. in-place)
+    row_means : 1D_array_like
+        Outpuit, Mean values of each row in `centered`
+    Returns
+    -------
+    global_mean : real
+        Global mean value
+    """
+    cdef Py_ssize_t n_samples = mat.shape[0]
+    cdef Py_ssize_t d2 = mat.shape[1]
+    cdef Py_ssize_t d3 = centered.shape[1]
+    cdef Py_ssize_t d4 = centered.shape[1]
+    cdef Py_ssize_t d5 = row_means.shape[0]
+
+    assert n_samples == d2
+    assert n_samples == d3
+    assert n_samples == d4
+    assert n_samples == d5
+
+    cdef Py_ssize_t row,col
+    cdef TReal row_sum
+    cdef TReal el0
+
+    cdef TReal global_sum = 0.0
+    for row in prange(n_samples, nogil=True):
+        row_sum = 0.0
+
+        for col in range(n_samples):
+            el0 = mat[row,col]
+            el0 =  -0.5*el0*el0
+            centered[row,col] = el0
+            # Note: do not use +=, so it is not flagged as a global reduction
+            row_sum = row_sum + el0
+
+        global_sum += row_sum
+        row_means[row] = row_sum/n_samples
+
+    return (global_sum/n_samples)/n_samples
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def f_matrix_inplace_cy(TReal[::1] row_means, TReal global_mean, TReal[:, ::1] centered):
+    """
+    Compute F matrix from E matrix inplace.
+    Centering step: for each element, the mean of the corresponding
+    row and column are subtracted, and the mean of the whole
+    matrix is added. Eq. 9.21 in Legendre & Legendre 1998.
+
+    Modified from :func:`skbio.stats.ordination.f_matrix_inplace` function,
+
+    Parameters
+    ----------
+    row_means : 1D_array_like
+        Mean values of each row in `centered`
+    global_mean : real
+        Global mean value in `centered`
+    centered : 2D array_like, must be symmetric
+        In,  a matrix representing the "E matrix" as described above.
+        Out, the centered matrix
+    """
+    cdef Py_ssize_t n_samples = centered.shape[0]
+    cdef Py_ssize_t d2 = centered.shape[1]
+    cdef Py_ssize_t d3 = row_means.shape[0]
+
+    assert n_samples == d2
+    assert n_samples == d3
+
+    cdef Py_ssize_t trow,tcol,row,col
+    cdef Py_ssize_t trow_max,tcol_max
+    cdef TReal gr_mean
+
+    # use a tiled pattern to maximize locality of row_means
+    for trow in prange(0, n_samples, 512, nogil=True):
+        trow_max = min(trow+512, n_samples)
+
+        for tcol in range(0, n_samples, 512):
+            tcol_max = min(tcol+512, n_samples)
+
+            for row in range(trow, trow_max, 1):
```

Comment:
Would it make sense to compute the array of `gr_mean` once? ie:

```python
gr_mean = np.array_like(row_means)  # or however to construct this
for row in prange(trow_max, nogil=True):
    gr_mean[row] = global_mean - row_means[row]
```

It seems like this is being recomputed through the nested loops, but I may be mislooking at something here

Reply:
it is much cheaper to do a simple subtraction at runtime instead of hosting another big memory buffer